Scrolling on Google Translate with keyboard up causes page to scroll back up

jesxilin opened the pull request:
#### 3574db4a22895aa5a005f60d321f67d9e73a2410
<pre>
Scrolling on Google Translate with keyboard up causes page to scroll back up
<a href="https://bugs.webkit.org/show_bug.cgi?id=293639">https://bugs.webkit.org/show_bug.cgi?id=293639</a>
<a href="https://rdar.apple.com/106539018">rdar://106539018</a>

Reviewed by Megan Gardner and Abrar Rahman Protyasha.

There is an issue where scrolling down on translate.google.com
with the keyboard up causes a jump back near the top of the page.
This is happening on iOS for Safari, as well as other browser apps.
(This issue is flakey on Chrome for iOS whereas the issue on other
browser apps, such as FireFox, is steadily reproducible.)

After web inspecting, it seems that the website is doing
something particular with calculating/changing `min-height` on the `body`
which could affect scrolling. When changing the user agent to Chrome
in the web inspector, the issue is no longer reproducible on Safari
or the other browser apps. In addition, the `min-height` value on
the `body` also no longer changes. In cases where the issue was reproducible,
the `min-height` value would change when the scrolling jump occurred.

So, write a quirk for changing the user agent to fix the
scrolling issue.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsGoogleTranslateScrollingQuirk const):
(WebCore::Quirks::scriptToEvaluateBeforeRunningScriptFromURL):
(WebCore::handleGoogleQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:

Canonical link: <a href="https://commits.webkit.org/295577@main">https://commits.webkit.org/295577@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b52eb64f32bf46a9c3e7d20c9aad7011d1a2e60

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105508 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25219 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15647 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110708 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/56158 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107549 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25689 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33760 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/80134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/56158 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108514 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20048 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95220 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60443 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/19773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13311 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55545 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/89502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13352 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113486 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32650 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/24076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/89209 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33013 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91450 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88869 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33740 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/11546 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28103 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17109 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32576 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/37984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32331 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35678 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33919 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lilyspiniolas commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -624,6 +624,15 @@ bool Quirks::needsGoogleMapsScrollingQuirk() const
+#if PLATFORM(IOS_FAMILY)
+    return needsQuirks() && m_quirksData.needsGoogleTranslateScrollingQuirk;
+#endif
+    return false;
```
```suggestion
#if PLATFORM(IOS_FAMILY)
    return needsQuirks() && m_quirksData.needsGoogleTranslateScrollingQuirk;
#else
    return false;
#endif
```

lilyspiniolas commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1573,6 +1582,9 @@ String Quirks::scriptToEvaluateBeforeRunningScriptFromURL(const URL& scriptURL)
     }
 
+    if (m_quirksData.needsGoogleTranslateScrollingQuirk && !scriptURL.isEmpty())
+        return "(function() { let userAgent = navigator.userAgent; Object.defineProperty(navigator, 'userAgent', { get: () => { return userAgent + ' Chrome/130.0.0.0 Android/15.0'; }, configurable: true }); })();"_s;
```
since this exact JS function is used elsewhere, it would be nice to have it as a static variable to prevent code duplication

pxlcoder commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1573,6 +1582,9 @@ String Quirks::scriptToEvaluateBeforeRunningScriptFromURL(const URL& scriptURL)
             return "(function() { let userAgent = navigator.userAgent; Object.defineProperty(navigator, 'userAgent', { get: () => { return userAgent + ' Chrome/130.0.0.0 Android/15.0'; }, configurable: true }); })();"_s;
     }
 
+    if (m_quirksData.needsGoogleTranslateScrollingQuirk && !scriptURL.isEmpty())
```
```suggestion
    if (m_quirksData.needsGoogleTranslateScrollingQuirk && !scriptURL.isEmpty()) [[unlikely]]
```

jesxilin committed:
Scrolling on Google Translate with keyboard up causes page to scroll back up
https://bugs.webkit.org/show_bug.cgi?id=293639
rdar://106539018

Reviewed by Megan Gardner and Abrar Rahman Protyasha.

There is an issue where scrolling down on translate.google.com
with the keyboard up causes a jump back near the top of the page.
This is happening on iOS for Safari, as well as other browser apps.
(This issue is flakey on Chrome for iOS whereas the issue on other
browser apps, such as FireFox, is steadily reproducible.)

After web inspecting, it seems that the website is doing
something particular with calculating/changing `min-height` on the `body`
which could affect scrolling. When changing the user agent to Chrome
in the web inspector, the issue is no longer reproducible on Safari
or the other browser apps. In addition, the `min-height` value on
the `body` also no longer changes. In cases where the issue was reproducible,
the `min-height` value would change when the scrolling jump occurred.

So, write a quirk for changing the user agent to fix the
scrolling issue.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsGoogleTranslateScrollingQuirk const):
(WebCore::Quirks::scriptToEvaluateBeforeRunningScriptFromURL):
(WebCore::handleGoogleQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:

Canonical link: https://commits.webkit.org/295577@main
